### PR TITLE
implemented storing wide search results in MongoDB

### PR DESCRIPTION
**Pull request by DFSko**
resolves happy-travel/agent-app-project#179

### CODE REVIEW

**Review by kirillta — APPROVED** [review]

File: Api/Controllers/AgentControllers/AvailabilitiesController.cs [hunk 1]
```diff
@@ -80,17 +81,22 @@ public async Task<IActionResult> GetAvailabilitySearchState([FromRoute] Guid sea
...
         [MinCounterpartyState(CounterpartyStates.ReadOnly)]
         [InAgencyPermissions(InAgencyPermissions.AccommodationAvailabilitySearch)]
-        [EnableQuery(MaxAnyAllExpressionDepth = 2, EnsureStableOrdering = false)]
-        public async Task<IEnumerable<WideAvailabilityResult>> GetAvailabilitySearchResult([FromRoute] Guid searchId)
+        public async Task<IActionResult> GetAvailabilitySearchResult([FromRoute] Guid searchId, ODataQueryOptions<WideAvailabilityResult> queryOptions)
         {
-            // TODO: Add validation and fool check for skip and top parameters
-            return await _wideAvailabilitySearchService.GetResult(searchId, await _agentContextService.GetAgent(), LanguageCode);
+            var top = queryOptions.Top?.Value;
+            var skip = queryOptions.Skip?.Value;
```
Comment (kirillta) [inline Api/Controllers/AgentControllers/AvailabilitiesController.cs:94]: Why can't we set defaults?

File: Api/Infrastructure/MongoDB/SerializationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public IBsonSerializer GetSerializer(Type type) => type switch
+        {
+            _ when type == typeof(AccommodationAvailabilityResult) => new AccommodationAvailabilityResultSerializer(),
+            _ when type == typeof(RoomContractSet) => new RoomContractSerializer(),
+            _ when type == typeof(Rate) => new RateSerializer(),
+            _ when type == typeof(MoneyAmount) => new MoneyAmountSerializer(),
+            _ when type == typeof(RoomContract) => new RoomContractSerializer(),
+            _ when type == typeof(DailyRate) => new DailyRateSerializer(),
+            _ => null
```
Comment (kirillta) [inline Api/Infrastructure/MongoDB/SerializationProvider.cs:19]: Not a default one for this case?

File: Api/Controllers/AgentControllers/AvailabilitiesController.cs (hunk 1 above)
Comment (kirillta) [reply Api/Controllers/AgentControllers/AvailabilitiesController.cs:94]: Also, a value is basically `queryOptions.Top?.Value?.Value` that's kind of awkward to split it to different calls (one here and another one below)

**Review by oleg1std — APPROVED** [review]

File: Api/Infrastructure/MongoDB/AvailabilityStorage.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+using MongoDB.Driver;
+using MongoDB.Driver.Linq;
+
+namespace HappyTravel.Edo.Api.Infrastructure.MongoDB
+{
+    public class AvailabilityStorage : IMongoDbStorage<CachedAccommodationAvailabilityResult>
+    {
+        public AvailabilityStorage(IOptions<MongoDbOptions> options)
+        {
+            var client = new MongoClient(options.Value.ConnectionString);
```
Comment (mingazhev) [inline Api/Infrastructure/MongoDB/AvailabilityStorage.cs:16]: `Typically you only create one MongoClient instance for a given cluster and use it across your application. Creating multiple MongoClients will, however, still share the same pool of connections if and only if the connection strings are identical.`

why not registtering singleton in startup and injecting it?

File: Api/Infrastructure/MongoDB/AvailabilityStorage.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            var database = client.GetDatabase(options.Value.DatabaseName);
+
+            _collection = database.GetCollection<CachedAccommodationAvailabilityResult>(nameof(CachedAccommodationAvailabilityResult));
```
Comment (mingazhev) [inline Api/Infrastructure/MongoDB/AvailabilityStorage.cs:19]: So, constructor is too complex. Consider using factory to create the collection

File: Api/Infrastructure/MongoDB/SerializationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+using HappyTravel.Edo.Api.Infrastructure.MongoDB.Serializers;
+using HappyTravel.Edo.Api.Models.Accommodations;
+using HappyTravel.Money.Models;
+using MongoDB.Bson.Serialization;
+
+namespace HappyTravel.Edo.Api.Infrastructure.MongoDB
+{
+    public class SerializationProvider : IBsonSerializationProvider
+    {
+        public IBsonSerializer GetSerializer(Type type) => type switch
```
Comment (mingazhev) [inline Api/Infrastructure/MongoDB/SerializationProvider.cs:11]: Arrow on the next line

File: Api/Infrastructure/MongoDB/SerializationProvider.cs (hunk 2 above)
Comment (mingazhev) [reply Api/Infrastructure/MongoDB/SerializationProvider.cs:19]: If this is an error, need to throw.
If it is ok to return null there, please add a comment why (will default serialized be used)

File: Api/Models/Storage/CachedAccommodationAvailabilityResult.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+using System;
+using HappyTravel.Edo.Api.Models.Accommodations;
+using HappyTravel.SuppliersCatalog;
+using MongoDB.Bson;
+
+namespace HappyTravel.Edo.Api.Models.Storage
+{
+    public record CachedAccommodationAvailabilityResult
```
Comment (mingazhev) [inline Api/Models/Storage/CachedAccommodationAvailabilityResult.cs:8]: It is not cached if speak strictly. Looks more like a record or entity

File: Api/Infrastructure/MongoDB/SerializationProvider.cs (hunk 2 above)
Comment (DFSko) [reply Api/Infrastructure/MongoDB/SerializationProvider.cs]: Fixed

File: Api/Infrastructure/MongoDB/AvailabilityStorage.cs (hunk 3 above)
Comment (DFSko) [reply Api/Infrastructure/MongoDB/AvailabilityStorage.cs]: Fixed

File: Api/Infrastructure/MongoDB/SerializationProvider.cs (hunk 5 above)
Comment (DFSko) [reply Api/Infrastructure/MongoDB/SerializationProvider.cs]: Fixed

File: Api/Controllers/AgentControllers/AvailabilitiesController.cs [hunk 7]
```diff
@@ -80,17 +81,22 @@ public async Task<IActionResult> GetAvailabilitySearchState([FromRoute] Guid sea
...
-        public async Task<IEnumerable<WideAvailabilityResult>> GetAvailabilitySearchResult([FromRoute] Guid searchId)
+        public async Task<IActionResult> GetAvailabilitySearchResult([FromRoute] Guid searchId, ODataQueryOptions<WideAvailabilityResult> queryOptions)
         {
-            // TODO: Add validation and fool check for skip and top parameters
-            return await _wideAvailabilitySearchService.GetResult(searchId, await _agentContextService.GetAgent(), LanguageCode);
+            var top = queryOptions.Top;
+            var skip = queryOptions.Skip;
+
+            if (top is null || skip is null)
+                return BadRequest("Pagination parameter `top` or `skip` not found");
```
[the remark this answers is not in the document]
Comment (DFSko) [reply Api/Controllers/AgentControllers/AvailabilitiesController.cs]: Fixed

File: Api/Infrastructure/MongoDB/AvailabilityStorage.cs (hunk 4 above)
Comment (DFSko) [reply Api/Infrastructure/MongoDB/AvailabilityStorage.cs]: Fixed

File: Api/Models/Storage/CachedAccommodationAvailabilityResult.cs (hunk 6 above)
Comment (DFSko) [reply Api/Models/Storage/CachedAccommodationAvailabilityResult.cs]: Fixed

**Comment by mingazhev** [comment]
PR was divided to multiple, closing

**Pull request closed without merging**